core/src/main/scala/kafka/log/ProducerIdMapping.scala

```diff
@@ -189,22 +193,36 @@ class ProducerIdMapping(val topicPartition: TopicPartition,
     }
   }
 
+  def checkForExpiredPids(currentTimeMs: Long) {
+    pidMap.retain { case (pid, lastEntry) =>
+      currentTimeMs - lastEntry.timestamp < maxPidExpirationMs
+    }
+  }
+
   def truncateAndReload(logEndOffset: Long) = {
     truncateSnapshotFiles(logEndOffset)
     loadFromSnapshot(logEndOffset)
   }
 
   /**
    * Update the mapping to the given epoch and sequence number.
-   * @param pid
-   * @param seq
-   * @param epoch
-   * @param offset
    */
-  def update(pid: Long, seq: Int, epoch: Short, offset: Long) {
-    if (pid > 0) {
-      pidMap.put(pid, PidEntry(seq, epoch, offset))
-      lastMapOffset = offset
+  def update(pid: Long, pidEntry: PidEntry) {
+    if (pid >= 0) {
```

Comment:
It should not be possible. We use -1 as the sentinel to indicate an invalid value. I debated raising an `IllegalArgumentException` if the pid was less than zero here.